tests/extensions/test_version.py

```diff
@@ -0,0 +1,210 @@
+
+    def tearDown(self):
+        super().tearDown()
+        self.item.validate()
+
+    def testStacExtensions(self):
+        self.assertEqual([pystac.Extensions.VERSION], self.item.stac_extensions)
+        # Make sure the item is valid for tearDown check.
+        self.item.ext.version.apply(self.version)
+
+    def testAddVersion(self):
+        self.item.ext.version.apply(self.version)
+        self.assertEqual(self.version, self.item.ext.version.version)
+        self.assertNotIn('deprecated', self.item.properties)
+        self.assertFalse(self.item.ext.version.deprecated)
+
+    def testVersionInProperties(self):
+        self.item.ext.version.apply(self.version, deprecated=True)
+        self.assertIn('version', self.item.properties)
+        self.assertIn('deprecated', self.item.properties)
+
+    def testAddNotDeprecatedVersion(self):
+        self.item.ext.version.apply(self.version, deprecated=False)
+        self.assertIn('deprecated', self.item.properties)
+        self.assertFalse(self.item.ext.version.deprecated)
+
+    def testAddDeprecatedVersion(self):
+        self.item.ext.version.apply(self.version, deprecated=True)
+        self.assertIn('deprecated', self.item.properties)
+        self.assertTrue(self.item.ext.version.deprecated)
+
+    def testLatest(self):
+        year = 2013
+        latest = MakeItem(year)
+        self.item.ext.version.apply(self.version, latest=latest)
+        latest_link = self.item.ext.version.latest_link
+        self.assertEqual(version.LATEST_VERSION, latest_link.rel)
+
+        expected_href = URL_TEMPLATE % year
+        self.assertEqual(expected_href, latest_link.get_href())
+
+    def testPredecessor(self):
+        year = 2010
+        predecessor = MakeItem(year)
+        self.item.ext.version.apply(self.version, predecessor=predecessor)
+        predecessor_link = self.item.ext.version.predecessor_link
+        self.assertEqual(version.PREDECESSOR_VERSION, predecessor_link.rel)
+
+        expected_href = URL_TEMPLATE % year
+        self.assertEqual(expected_href, predecessor_link.get_href())
+
+    def testSuccessor(self):
+        year = 2012
+        successor = MakeItem(year)
+        self.item.ext.version.apply(self.version, successor=successor)
+        successor_link = self.item.ext.version.successor_link
+        self.assertEqual(version.SUCCESSOR_VERSION, successor_link.rel)
+
+        expected_href = URL_TEMPLATE % year
+        self.assertEqual(expected_href, successor_link.get_href())
+
+    def testFailValidate(self):
+        with self.assertRaises(pystac.validation.STACValidationError):
+            self.item.validate()
+
+        self.item.ext.version.apply(self.version)
+
+    def testAllLinks(self):
+        deprecated = True
+        latest = MakeItem(2013)
+        predecessor = MakeItem(2010)
+        successor = MakeItem(2012)
+        self.item.ext.version.apply(self.version, deprecated, latest, predecessor, successor)
+        # Just validate in tearDown.
```

Comment:
What I have so far:

```
    def test_copy_1_with_version(self):
        c2011 = make_collection(2011)
        c2012 = make_collection(2012)
        c2013 = make_collection(2013)
        c2014 = make_collection(2014)

        c2012.ext.enable(pystac.Extensions.VERSION)
        c2012.ext.version.apply('v2012', latest=c2012, predecessor=c2011, successor=c2013)

        cat = pystac.Catalog('my_catalog', 'my description')
        cat.add_child(c2011)
        cat.add_child(c2012)
        cat.add_child(c2013)
        cat.add_child(c2014)

        cat_copy = cat.full_copy()
```

Was trying to figure out how best to compare the two.  Hopefully I can think through it easier where there aren't mindcraft videos playing next to me... :|

Reply:
Having a hard time figuring out the item link thing.  This is what I had as of the end of yesterday

```#!python
def compare_collection(col1, col2):
  if col1.id != col2.id:
    return False
  if col1.description != col2.description:
    return False
  has_ver_1 = col1.ext.implements(pystac.Extensions.VERSION)
  has_ver_2 = col2.ext.implements(pystac.Extensions.VERSION)
  if has_ver_1 != has_ver_2:
    return False
  if has_ver_1:
    print('has_ver')
    ver1 = col1.ext.version
    ver2 = col2.ext.version
    if ver1.version != ver2.version:
      return False
    if ver1.deprecated != ver2.deprecated:
      return False
    if str(ver1.latest_link) != str(ver2.latest_link):
      return False
    if str(ver1.predecessor_link) != str(ver2.predecessor_link):
      return False
    if str(ver1.successor_link) != str(ver2.successor_link):
      return False
  return True


class FullCopyTest(googletest.TestCase):

  def compare_catalogs(self, cat1, cat2):
    self.assertEqual(cat1.id, cat2.id)
    children1 = {child.id: child for child in cat1.get_children()}
    # print('children1 dict', children1)
    children2 = {child.id: child for child in cat2.get_children()}
    # print('children2 dict', children2)
    self.assertCountEqual(sorted(children1.keys()), sorted(children2.keys()))
    for child_id in children1:
      print(child_id)
      child1 = children1[child_id]
      child2 = children2[child_id]
      has_ver_1 = child1.implements(pystac.Extensions.VERSION)
      has_ver_2 = child1.implements(pystac.Extensions.VERSION)
      self.assertEqual(has_ver_1, has_ver_2)
      if has_ver_1:
        # Umm... ?
        self.assertTrue(compare_collection(child1, child2))
        link = child1.get_links(version.LATEST_VERSION)[0].resolve_stac_object()
        self.assertIs(link.target, child2)

    return True

  def testCopy1Linked(self):
    c2011 = make_collection(2011)
    c2012 = make_collection(2012)
    c2013 = make_collection(2013)
    c2014 = make_collection(2014)

    c2012.ext.enable(pystac.Extensions.VERSION)
    c2012.ext.version.apply(
        'v2012', latest=c2014, predecessor=c2011, successor=c2013)

    cat = pystac.Catalog('my_catalog', 'my description')
    cat.add_child(c2011)
    cat.add_child(c2012)
    cat.add_child(c2013)
    cat.add_child(c2014)

    cat_copy = cat.full_copy()
    pprint(cat_copy.to_dict())

    self.compare_catalogs(cat, cat_copy)
```

Reply:
The easier thing would be to do a straight object identity test to ensure the object cache ensures there's a single object that's linked to in the catalog and by the item that has the version link.

It might be easier/more clear to make the suggestion in code. I can take a crack at this tomorrow if you're alright with me pushing a commit to this branch?

Reply:
Go for it!  The only things I really care about is learning how it's done and getting the extensions implemented.  I don't like what I did above except for the setup (assuming that it makes input of the form required.)